src/nvim/regexp_nfa.c

```diff
@@ -3535,11 +3537,13 @@ static void copy_ze_off(regsub_T *to, regsub_T *from)
 {
   if (nfa_has_zend) {
     if (REG_MULTI) {
-      if (from->list.multi[0].end.lnum >= 0)
-        to->list.multi[0].end = from->list.multi[0].end;
-    } else {
-      if (from->list.line[0].end != NULL)
-        to->list.line[0].end = from->list.line[0].end;
+      if (from->list.multi[0].end_lnum >= 0){
+        to->list.multi[0].end_lnum = from->list.multi[0].end_lnum;
+        to->list.multi[0].end_col = from->list.multi[0].end_col;
+      } else {
+        if (from->list.line[0].end != NULL)
+          to->list.line[0].end = from->list.line[0].end;
+      }
```

Comment:
It's a bit hard to see, but you seem to have moved the else-case of the `if(REG_MULTI)` case inside it, which I think the original patch doesn't do.